### ISSUE TITLE
Miners consume a lot of food

### ISSUE THREAD

**Issue by nightyhawk** [Pyrdacor/freeserf.net#115]
At first - great development so far.

During some playing sessions I realized, that a miner will take almost one food for every raw material he is digging, 

In the original game there is a small chance, that he starts without any food for one material. But after being feeded one piece of food last for some more materials (coult not state how much - but he will get approximately up to 10 pieces of raw material for every food piece).

**Comment by Pyrdacor** [comment]
Hi thanks for your feedback. I guess the chances are a bit off then. Could be hard to find the right values but I will look into it.

**Comment by Pyrdacor** [comment]
TL;DR: Ok I digged into it and found out that miners indeed consume food every time. There is only a small chance that they don't. So when they consumed food there is no increased chance of consuming no food next time.

I added a motivation value. It wasn't easy to implement as serf behavior is controlled by a sequence of states and passing values between them isn't that easy. So if a miner gets a resource he will go through several states (move resource out -> leave building -> drop resource out -> ready to enter -> entering building -> next mining). Passing the motivation wasn't that trivial but I managed it. So it should work now.

The values are totally made up by me:

- Initial the motivation is 0 but the small chance is still there that no food is needed
- If the miner consumes food the motivation is set to 128 + a random number between 0 and 127 (so 128-255 in total)
- If no resource was found, motivation decreases by 63
- If a resource was found, motivation decreases by 31

Everytime the miner starts to work a random number between 0 and 255 is generated. If the number is smaller than motivation, no food is consumed.

-----------------------------------

Should be fixed/improved in v2.0.7.

**Comment by nightyhawk** [comment]
Hey,
I had a look over your changeset and it looked nice. I'm not sure how close it is compared to the original game mechanics - but in any case it is better than before.

I wonder, if the line 
_if ((random & 7) == 0)_
is still needed. In my opinion this holds only for the first entrance into the mine. Afterwards one should only look to new implemented dice roll - but I'm not sure, how this can be designed in your state machine.

Btw - I build a coal mine yesterday in original "Die Siedler". I played it that way, that the mine was not supplied with any food - nevertheless my brave miner started seven attemps to dig for some coal.
Afterwards I let one piece of food in, which lead to two more attempts to dig.
Again one piece of food - and two more attempts.

And now it gets strange -> I now let the mine be fullfilled with 5 pieces of food (one consumed immediatley) and 4 pieces into storage - this holds for 21 attempts until the next food was eaten.
After that the miner digged 19 times without eating. Quite impressive numbers....

**Comment by Pyrdacor** [comment]
I kept the small random to add some variation, eased code changes and allow to extend the max number of possible digging.

But your recent findings suggest, that the motivation decrease depends on total food in the mine. Could you test a bit more on this? I am very thankful for your testing in original game. :)

**Comment by Pyrdacor** [comment]
Maybe motivation is only decreased when no resource was found? Or there is just a "number of failed attempts" counter which is also set to 1 with a small chance on first try? This could explain the 7 diggings without food. But only if 6 resources were found before.

In my memory only 1 or 2 minerals were found without food. But I could be wrong.

**Comment by nightyhawk** [comment]
Well - I'll try to have some rounds of "mine digging" tomorrow... maybe we can come closer to the original numbers....

**Comment by Pyrdacor** [comment]
Thanks

**Comment by nicymike** [comment]
Hey, just switch substates (first = 1, second = 2) which will fix the issue. I did check the original source to validate this.
```
/* There is a small chance that the miner will
   require food and skip to state 2. */
int random = Game.RandomInt();
if ((random & 7) == 0)
{
	stateData.Mining.Substate = 1;
}
else
{
	stateData.Mining.Substate = 2;
}
```

**Comment by nightyhawk** [comment]
Oh right - I promised to mine some materials for you... Unfortunately I had not much time to record some nice statistics - nevertheless I have the feeling, that one piece of food lasts for a long time - at least for ten up to twenty rounds of digging before it is consumed......

Maybe @nicymike good give some advice, how long it takes till one piece of food is consumed?

**Comment by nicymike** [comment]
@nightyhawk 
The game generates a random number, does a bitwise AND calculation with seven (random & 111) and if the result is zero the miner needs one food for the next mine. It all depends on random.

**Comment by Pyrdacor** [comment]
So the chance is 1/8 to consume food. Which is 12,5%. I always had the feeling that without food a miner would only sometimes work 1 or at max 2 times.

If this is only the 1/8 random check I can adjust the code accordingly.

**Comment by Pyrdacor** [comment]
Ok I added the change and tested it. Looks very good. I reverted my other changes regarding food consumption. I released the change as v2.0.7. I don't have much time for freeserf.net at the moment but I wanted at least to add the fix to improve the play experience.

@nightyhawk I guess there is no need anymore for mining tests.
@nicymike Thanks for your solution.

**Comment by nicymike** [comment]
@Pyrdacor No worries

**Comment by tlongstretch** [comment]
it seems this was originally correct in Freeserf but in 2013 jonls flipped it to 7/8 consumption because he felt 1/8 was too easy.  And I agree it is too easy, but I prefer it to default to the correct value!  that is a huge change
